pkg/queryfrontend/cache_splitter.go

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package queryfrontend
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+)
+
+// constSplitter is a utility for using a constant split interval when determining cache keys.
+type constSplitter time.Duration
+
+// TODO(yeya24): Add other request params as request key.
+// GenerateCacheKey generates a cache key based on the Request and interval.
+func (t constSplitter) GenerateCacheKey(_ string, r queryrange.Request) string {
+	currentInterval := r.GetStart() / time.Duration(t).Milliseconds()
+	if tr, ok := r.(*ThanosRequest); ok {
+		return fmt.Sprintf("%s:%d:%d:%d", tr.Query, tr.Step, currentInterval, tr.MaxSourceResolution)
```

Comment:
Hmmmm I would be careful here. E.g maxSourceResolution `1` and `2` will have exactly the same result, so no need to cache it differently.

Happy to start with this and improve later though.

Worth to add TODO here?